Improve Path2D debug performance

smix8 opened the pull request:
Improves `Path2D` node debug performance by changing the debug to use mesh and multimesh instead of the slower canvas line draw functions.

This PR removes the old remnants of what was a fake open-gl line anti-aliasing that no longer works (it rendered everything twice with 0.5 alpha, aka very efficient).

The PR also removes the debug path width that did not really work and caused a lot of rendering issues on high res and dpi displays or different zoom and scale levels in the editor. It is now all thin "hardware" mesh lines.

![linezoom](https://github.com/user-attachments/assets/ad912757-cd03-4f1f-a19d-8e224227be85)

This largely means it replaces the old debug rendering. For editing paths the internal arrays got replaced with more efficient LocalVectors that keep the memory between path edits.

The rather slow canvas item draw for each segment and point were replaced with a combination of using static meshes for path lines and several multimeshes for the rendering of the editor handles and textures.

The nodes now always render the static debug visuals while the editor plugin only renders the editing parts like handles of the selected node on top.

![thick](https://github.com/user-attachments/assets/11a065c4-10f3-4ee2-b2a5-6a907828cb30)

The result is a very noticeable performance boost editing and rendering many complex 2d paths at the same time.
<!--
Please target the `master` branch in priority.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->

AThousandShips commented on scene/2d/path_2d.cpp:
```diff
@@ -30,13 +30,137 @@
+void PathDebug2D::emit_update_callbacks() {
+	MutexLock lock(update_callbacks_mutex);
+	for (KeyValue<Path2D *, Callable> &kv : update_callbacks) {
+		ERR_CONTINUE(kv.value.is_null());
```
```suggestion
		ERR_CONTINUE(!kv.value.is_valid());
```
`Callable` should be checked against `is_valid` for checking if it can be called

AThousandShips commented on scene/3d/path_3d.cpp:
```diff
@@ -30,24 +30,152 @@
+void PathDebug3D::emit_update_callbacks() {
+	MutexLock lock(update_callbacks_mutex);
+	for (KeyValue<Path3D *, Callable> &kv : update_callbacks) {
+		ERR_CONTINUE(kv.value.is_null());
```
```suggestion
		ERR_CONTINUE(!kv.value.is_valid());
```

Mickeon commented:
I personally feel like the extent in which this PR goes to customize debug drawing is quite excessive. Especially the custom sampling. It looks like something that the editor could infer automatically based on current performance, distance from the cure, overall length, and so on.

Does this PR happen to also supersede https://github.com/godotengine/godot/pull/90357 ?

smix8 commented:
@Mickeon
The editor can not infer this as long as it is not living in the users head. This is not about performance only but also user intention, so it needs settings. E.g. multiple times different people tried to convince me that a few more fish bones, a few less bones, or even so many fish bones that you can not spot the line anymore would be good and appropriated. Then we had people that said they needed the detail for only some important paths, and wanted all the less important to be coarse, some said they did not care at all and wanted it off, ... oh, and only for some paths btw. With such vastly different views and uses among just a handful of people how would you automate this? it is not possible.

Calinou commented:
Can the curve sampling be automatically determined based on whether the current/previous/next points use any curve tangents? If it's a straight line, it can be drawn with a single vertex. This would also optimize line drawing for lines that only have some curved points (but not all).

Also, I think **Custom Color** should be moved at the top of the custom options (below **Custom Enabled**) as it's likely the one users will tweak the most.

smix8 commented:
> Can the curve sampling be automatically determined based on whether the current/previous/next points use any curve tangents? 

Yes oc you can preprocess everything but always the question if you gain more than you lose. This info if a segment is straight or curved should likely be part of the actual Curve resource point meta cache and not of the Path2D/3D Node as you don't want to recalculated it again for every single node that uses the curve.

Mickeon commented on doc/classes/Path2D.xml:
```diff
@@ -9,9 +9,55 @@
+			<return type="void" />
+			<param index="0" name="enabled" type="bool" />
+			<description>
+				If [code]true[/code], enables the global path debug.
```
Heavily needs a proper explanation, over "enables the global path debug". Also a link to its associated setter. Something along the lines of:
```suggestion
				If [code]true[/code], enables debugging for all [Path2D] in the scene tree. See also [method set_debug_global_enabled].
```
I cannot say for certain what this means, as of writing this.

Mickeon commented on doc/classes/Path2D.xml:
```diff
@@ -9,9 +9,55 @@
+		<method name="is_debug_global_enabled" qualifiers="const">
+			<return type="bool" />
+			<description>
+				Returns [code]true[/code] if the global path debug is enabled.
```
Thoughts are similar to a prior review.

Mickeon commented on doc/classes/Path2D.xml:
```diff
@@ -9,9 +9,55 @@
 		<member name="curve" type="Curve2D" setter="set_curve" getter="get_curve">
 			A [Curve2D] describing the path.
 		</member>
+		<member name="debug_custom_color" type="Color" setter="set_debug_custom_color" getter="get_debug_custom_color" default="Color(1, 1, 1, 1)">
+			The custom color used to draw the path debug when [member debug_custom_enabled] is enabled.
+		</member>
+		<member name="debug_custom_enabled" type="bool" setter="set_debug_custom_enabled" getter="get_debug_custom_enabled" default="false">
+			If [code]true[/code], the debug custom properties override the global defaults.
+		</member>
```
Note that a similar API in **CollisionShape2D** doesn't work like this. If `CollisionShape2D.debug_color` is `Color(0, 0, 0, 0)`, the value defined in the project settings is used. One could make similar arguments for the other properties, but this is the most striking.

smix8 replied:
Using any magic values to control debug is not a great because not only do you lay a trap for people that want to use that value but you force people to script book-keep that value when they want to toggle between custom and global debug.

Mickeon replied:
In the case of colour it seems completely fine, and Godot is no stranger to it. Some color theme properties behave the same.

Either way, what's important to me here is that a different implementation for a similar concept already exists. Whether you disagree with it (I sympathise), this inconsistency will be a bit of a head scratcher. **CollisionShape3D** also has `debug_color`.

Mickeon commented on doc/classes/Path2D.xml:
```diff
@@ -9,9 +9,55 @@
+			If [code]true[/code], the debug display adds additional fish bones along the path segments to visualize direction. Used when [member debug_custom_sampling_enabled] is enabled.
+		</member>
+		<member name="debug_custom_fish_bones_interval" type="int" setter="set_debug_custom_fish_bones_interval" getter="get_debug_custom_fish_bones_interval" default="4">
+			The custom interval for adding fish bones to the path sample points. E.g. a value of 4 places a fish bone for every 4-th sampled point. Used when [member debug_custom_sampling_enabled] is enabled.
```
The hyphen is not needed, correct me if it's an American English thing or the secret forbidden rule.
```suggestion
			The custom interval used to render fish bones to the points sampled in the path. For example, a value of [code]4[/code] renders a fish bone every 4th sampled point. Used when [member debug_custom_sampling_enabled] is [code]true[/code].
```

Mickeon commented on doc/classes/Path2D.xml:
```diff
@@ -9,9 +9,55 @@
+			The custom interval for adding fish bones to the path sample points. E.g. a value of 4 places a fish bone for every 4-th sampled point. Used when [member debug_custom_sampling_enabled] is enabled.
+		</member>
+		<member name="debug_custom_sample_interval" type="float" setter="set_debug_custom_sample_interval" getter="get_debug_custom_sample_interval" default="5.0">
+			The custom interval for sampling the curve for path sample points. Used when [member debug_custom_sampling_enabled] is enabled.
```
Assuming the following is correct:
```suggestion
			The custom interval used to sample the points in the path's curve. Used when [member debug_custom_sampling_enabled] is enabled.
```
Although this description, and this property's name, implies it overrides the original Curve's `bake_resolution`? Not sure.

Mickeon commented on scene/2d/path_2d.cpp:
```diff
@@ -201,7 +631,46 @@ void Path2D::_bind_methods() {
+	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "debug_custom_fish_bones_enabled"), "set_debug_custom_fish_bones_enabled", "get_debug_custom_fish_bones_enabled");
+	ADD_PROPERTY(PropertyInfo(Variant::INT, "debug_custom_fish_bones_interval", PROPERTY_HINT_RANGE, "1,10,1,or_greater"), "set_debug_custom_fish_bones_interval", "get_debug_custom_fish_bones_interval");
+
+	ADD_SIGNAL(MethodInfo("curve_changed"));
```
I get why it exists, but exposing a new signal to the API of one of the most commonly used nodes in the engine... It warrants its own PR, in my opinion. Tread carefully.

smix8 replied:
It exists on Path3D but I can just remove it.

Path2D and Path3D are half the time in a messy desync anyway so the damage is already done for their api.

Mickeon reviewed (commented):
Still very much in the opinion that all of these. I think we seriously need to be careful about property bloat. Please let's tread carefully.

At the very least, these should be surrounded with **DEBUG_ENABLED** (if they're not already)

The documentation review is not exhaustive but I'm getting just a few out of the way.
It feels like the descriptions were written by programmers, for programmers (and, I mean, of course they were). It's a bit hard to explain how, but it is especially problematic because these properties are prominently present in the Inspector and potentially changed by Godot developers (such as level designers), not necessarily users that know how to code.

smix8 commented on doc/classes/Path2D.xml:
```diff
@@ -9,9 +9,55 @@
 	<tutorials>
 	</tutorials>
+	<methods>
+		<method name="is_debug_global_enabled" qualifiers="const">
```
Making this function static does not work for the script binds.

Mickeon replied:
May I ask how? If this doesn't affect all Path2D objects, could this be a misnomer?

KoBeWi replied:
You need to use `bind_static_method()` for static methods.

smix8 replied:
It affects all path nodes as they register with their callback for changes. It just does not work as a static function because the ClassDB can not handle it.

smix8 commented:
Removed most of the customization options and functions on the nodes except color and the toggle for runtime debug.

Mickeon commented on doc/classes/ProjectSettings.xml:
```diff
@@ -868,11 +868,29 @@
-		<member name="debug/shapes/paths/geometry_width" type="float" setter="" getter="" default="2.0">
-			Line width of the curve path geometry, visible when "Visible Paths" is enabled in the Debug menu.
+		<member name="debug/shapes/paths/2d/fish_bones_interval" type="int" setter="" getter="" default="4">
+			The default [Path2D] curve path sample point interval in which fish bones are places along the path. E.g. a value of 4 places a fish bone for every 4-th sampled point along path.
```
Similar to https://github.com/godotengine/godot/pull/105606#discussion_r2052882421 .


```suggestion
			The default interval used to display fish bones to the points sampled in a path, for [Path2D] nodes. For example, a value of [code]4[/code] displays a fish bone every 4th sampled point along the path. Used when [member debug/shapes/paths/2d/fish_bones_enabled] is [code]true[/code] and "Visible Paths" is enabled in the Debug menu.
```

"_to display fish bones_" instead of "_to render fish bones_" because... I don't know, it makes more sense and allows the description to be the same for 3D?

The reference to the Debug menu is for consistency with the rest.

_"curve path sample point interval"_ is quite the mouthful. Imagine translating it to another language. Interval of sample point of path of the curve? Interval of a point of the sample of the path... of the curve?
I am saying this for future reference.

Mickeon commented on doc/classes/ProjectSettings.xml:
```diff
@@ -868,11 +868,29 @@
+			Color of the [Path2D] curve path geometry, visible when "Visible Paths" is enabled in the Debug menu.
+		</member>
+		<member name="debug/shapes/paths/2d/sample_interval" type="float" setter="" getter="" default="5.0">
+			The default [Path2D] curve path sample interval in which points are sampled for debug display.
```
```suggestion
			The interval used to sample the points in a path's curve, for [Path2D] nodes. Used when "Visible Paths" is enabled in the Debug menu.
```

Mickeon commented on doc/classes/ProjectSettings.xml:
```diff
@@ -868,11 +868,29 @@
+			The default [Path2D] curve path sample point interval in which fish bones are places along the path. E.g. a value of 4 places a fish bone for every 4-th sampled point along path.
+		</member>
+		<member name="debug/shapes/paths/2d/geometry_color" type="Color" setter="" getter="" default="Color(0.1, 1, 0.7, 0.4)">
+			Color of the [Path2D] curve path geometry, visible when "Visible Paths" is enabled in the Debug menu.
```
Something like this? I'm not sure it's accurate, or maybe it's too many generic "_used_":
```suggestion
			The color used to display the curve of a path, for [Path2D] nodes. Used when "Visible Paths" is enabled in the Debug menu.
```
or
```suggestion
			The color used to display a path, for [Path2D] nodes. Used when "Visible Paths" is enabled in the Debug menu.
```
or similar?

Mickeon commented on doc/classes/Path2D.xml:
```diff
@@ -13,5 +13,14 @@
+			If [code]true[/code], the debug custom properties override the global defaults.
+		</member>
+		<member name="debug_enabled" type="bool" setter="set_debug_enabled" getter="get_debug_enabled" default="true">
+			Used at runtime to toggle debug when "Visible Paths" is enabled in the Editor Debug menu. Inside the Editor debug is always enabled.
```
This description only works with context. Without reading the property's name and type, you'd not understand it.
I think this is related to `debug_paths_hint`?
```suggestion
			If [code]true[/code], allows this path to be displayed for debugging purposes. Used when "Visible Paths" is enabled in the Editor Debug menu. In the editor, this property has no effect. See also [member SceneTree.debug_paths_hint].
```

Mickeon commented on doc/classes/Path2D.xml:
```diff
@@ -13,5 +13,14 @@
+		<member name="debug_custom_enabled" type="bool" setter="set_debug_custom_enabled" getter="get_debug_custom_enabled" default="false">
+			If [code]true[/code], the debug custom properties override the global defaults.
+		</member>
+		<member name="debug_enabled" type="bool" setter="set_debug_enabled" getter="get_debug_enabled" default="true">
```
Isn't exposing `debug_enabled` a bit redundant, since you could set `debug_custom_color` to a fully transparent colour to essentially get rid of the visualization?

Mickeon reviewed (commented):
The following suggestions were done to the 2D versions of the descriptions, but they would also apply to the 3D equivalents with appropriate changes.
It's not necessarily exhaustive, but I'm waiting for others to chime in, too.

I cannot speak much about this, code-wise. Just that the colour for every path is the most useful customization provided, as I already hinted at.

My thoughts in https://github.com/godotengine/godot/pull/105606#discussion_r2052875614 still stand, further compounded by there being only one "custom" property now.

KoBeWi commented:
Error spam when hovering or editing a Path2D: `servers\rendering_server.cpp:1193 - Condition "array_len == 0" is true. Returning: ERR_INVALID_DATA`

The new project settings require editor restart to get applied. They should be either marked as requiring restart or properly handle changing settings (at least in the editor). You can use the new `GLOBAL_GET_CACHED` for that.

KoBeWi commented on scene/2d/path_2d.h:
```diff
@@ -33,15 +33,75 @@
 class Timer;
 
+class PathDebug2D {
+	static bool debug_enabled;
+	static Color debug_paths_color;
+	static float debug_paths_width;
+	static float debug_paths_sample_interval;
+	static bool debug_paths_fish_bones_enabled;
+	static int debug_paths_fish_bones_interval;
```
Static variables should be inlined.

KoBeWi commented on scene/2d/path_2d.cpp:
```diff
@@ -201,7 +479,21 @@ void Path2D::_bind_methods() {
+
+	ADD_GROUP("Debug", "debug_");
+	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "debug_enabled"), "set_debug_enabled", "get_debug_enabled");
+	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "debug_custom_enabled"), "set_debug_custom_enabled", "get_debug_custom_enabled");
```
This property only controls the color. Why is it needed? You can use transparent color as default.

Mickeon replied:
Also https://github.com/godotengine/godot/pull/105606/files#r2052875614

KoBeWi commented on scene/3d/path_3d.h:
```diff
@@ -33,22 +33,80 @@
+	static Ref<Material> debug_default_material;
+
+	static void emit_changed();
+	static bool _emitting_changed;
```
Unused.

KoBeWi commented on scene/3d/path_3d.h:
```diff
@@ -33,22 +33,80 @@
+	static int debug_paths_fish_bones_interval;
+	static Ref<Material> debug_default_material;
+
+	static void emit_changed();
+	static bool _emitting_changed;
+	void _emit_changed_deferred();
```
Unused (at least the bottom 2).

KoBeWi commented on scene/2d/path_2d.h:
```diff
@@ -33,15 +33,75 @@
+	static void set_debug_paths_fish_bones_interval(int p_interval);
+	static int get_debug_paths_fish_bones_interval();
+
+	PathDebug2D();
+	~PathDebug2D();
```
```suggestion
```
Not implemented.

KoBeWi commented on scene/2d/path_2d.cpp:
```diff
@@ -201,7 +479,21 @@ void Path2D::_bind_methods() {
 	ADD_PROPERTY(PropertyInfo(Variant::OBJECT, "curve", PROPERTY_HINT_RESOURCE_TYPE, "Curve2D", PROPERTY_USAGE_DEFAULT | PROPERTY_USAGE_EDITOR_INSTANTIATE_OBJECT), "set_curve", "get_curve");
+
+	ADD_GROUP("Debug", "debug_");
+	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "debug_enabled"), "set_debug_enabled", "get_debug_enabled");
```
This property is misleading. It does nothing in the editor and does nothing at runtime, unless you enable Path Debug.
While this is properly documented, it might be better to rename it to `disable_debug` or something similar, to make its only use clearer.

KoBeWi commented on scene/2d/path_2d.cpp:
```diff
@@ -87,81 +215,231 @@ bool Path2D::_edit_is_selected_on_click(const Point2 &p_point, double p_toleranc
+		case NOTIFICATION_ENTER_TREE: {
+			PathDebug2D::add_update_callback(this, callable_mp(this, &Path2D::_on_debug_global_changed));
+			_debug_create();
+			queue_redraw();
```
Redrawing is done automatically when entering tree.

KoBeWi commented on scene/2d/path_2d.cpp:
```diff
@@ -87,81 +215,231 @@ bool Path2D::_edit_is_selected_on_click(const Point2 &p_point, double p_toleranc
 	switch (p_what) {
-		// Draw the curve if path debugging is enabled.
+		case NOTIFICATION_ENTER_TREE: {
+			PathDebug2D::add_update_callback(this, callable_mp(this, &Path2D::_on_debug_global_changed));
+			_debug_create();
```
This should be wrapped in DEBUG_ENABLED.

KoBeWi commented on scene/2d/path_2d.h:
```diff
@@ -33,15 +33,75 @@
+	void _debug_clear();
+	void _debug_free();
+
+	bool _emitting_debug_changed = false;
```
Why is this underscored?
Also could use a better name, like `debug_emit_queued`.

KoBeWi commented on main/main.cpp:
```diff
@@ -4167,7 +4173,8 @@ int Main::start() {
 		if (debug_paths) {
-			sml->set_debug_paths_hint(true);
+			PathDebug2D::set_debug_enabled(true);
+			PathDebug3D::set_debug_enabled(true);
```
This is not wrapped in `_3D_DISABLED`, so it likely won't compile.

KoBeWi commented on scene/2d/path_2d.cpp:
```diff
@@ -87,81 +215,231 @@ bool Path2D::_edit_is_selected_on_click(const Point2 &p_point, double p_toleranc
-			if (curve->get_point_count() < 2) {
-				return;
-			}
+void Path2D::_emit_debug_changed_deferred() {
```
This method's name is also super misleading. It does not emit anything.

smix8 replied:
agree, it is remnant. That function originally did something else hence the naming.

KoBeWi commented on scene/2d/path_2d.cpp:
```diff
@@ -87,81 +215,231 @@ bool Path2D::_edit_is_selected_on_click(const Point2 &p_point, double p_toleranc
-#endif
-			real_t interval = 10;
-			const real_t length = curve->get_baked_length();
+	if (is_inside_tree()) {
```
`queue_redraw()` already checks this.

KoBeWi commented on scene/2d/path_2d.cpp:
```diff
@@ -87,81 +215,231 @@ bool Path2D::_edit_is_selected_on_click(const Point2 &p_point, double p_toleranc
+		bones.resize(sample_count * vertex_per_bone);
+		Vector2 *bones_ptrw = bones.ptrw();
+
+		for (int i = 0; i < sample_count / fish_bones_interval; i++) {
```
`sample_count / fish_bones_interval` will get recalculated every time. You can do it once, before the loop (unless it can change).

smix8 committed:
Improve Path2D debug performance

Improves Path2D node debug performance by changing the debug to use mesh and multimesh instead of slower canvas draw functions.

smix8 commented:
> The error spam is still not fixed.

I think I found the issue and fixed it now. At least I was not able to trigger such error again.

KoBeWi reviewed (approved):
I did not compare performance, but it works correctly.

Repiteo commented:
Thanks!
